Add mechanism for transporting shard-level actions by node

jasontedor opened the pull request:
Currently, many shard-level operations are transported with a request
per shard via TransportBroadcastAction. These shard-level requests are
then submitted to unbounded execution queues for asynchronous execution
on the receiving node. This transport mechanism and stuffing of the
execution queues can be problematic on large clusters. A better
mechanism would be to aggregate the shard-level requests, transport
them via a single request per node, and execute the shard-level
operations serially on the receiving node.

This commit introduces TransportNodeBroadcastAction which is the
high-level mechanism for transporting the shard-level operations in a
single request per node. The shard-level operations are executed
serially on the receiving node and per-node shard-level results are
aggregated into a single response per node. These node-level results
are then aggregated into a single response to the initial request.

One item of note is a new mechanism for registering request handlers.
This mechanism enables registrants to provide a callback for
instantiating new instances of the request class. Doing this enables
the inner class to be instantiated with the context of its outer class.
This is done so that a single NodeRequest class can be defined rather
than defining a class per operation.

Closes #7990

nik9000 commented on core/src/main/java/org/elasticsearch/action/support/indices/TransportNodeBroadcastAction.java:
```diff
@@ -0,0 +1,309 @@
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReferenceArray;
+
+public abstract class TransportNodeBroadcastAction<Request extends IndicesLevelRequest,
```
Javadocs would be super helpful here.

jpountz commented:
I'm slightly worried that we are optimizing for a use-case (too many shards/indices) that we also happen to recommend against? Besides I'm wondering how much it will help: even if we transport a single request instead of N, I don't see it helping save memory (https://github.com/elastic/elasticsearch/issues/9130) or make stats calls faster (https://github.com/elastic/elasticsearch/issues/7385) which seem to be motivations for #7990?

nik9000 commented:
> I'm slightly worried that we are optimizing for a use-case (too many shards/indices) that we also happen to recommend against?

We recommend against oversubscription as a way of optimizing write throughput but we recommend a little oversubscription if you expect to add more nodes. We also commonly recommend time based indexes which can lead to lots of situations like this.

> Besides I'm wondering how much it will help: even if we transport a single request instead of N, I don't see it helping save memory (#9130) or make stats calls faster (#7385) which seem to be motivations for #7990?

I haven't done enough research to refute or confirm these claims.

jasontedor commented:
@jpountz The initial version of this pull request contained only the change to send a single request to each node rather than a request per shard. A subsequent [commit](https://github.com/jasontedor/elasticsearch/commit/0d2c91663243abfffff6bc5a0c11761f6714f035) has updated this pull request to also execute the shard operations in serial on the receiving node. The advantage of this is to avoid filling up unbounded execution queues (e.g., the management and optimize queues) with many shard-level operations on large clusters.

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/cache/clear/TransportClearIndicesCacheAction.java:
```diff
@@ -65,42 +61,30 @@ public TransportClearIndicesCacheAction(Settings settings, ThreadPool threadPool
     }
 
     @Override
-    protected ClearIndicesCacheResponse newResponse(ClearIndicesCacheRequest request, AtomicReferenceArray shardsResponses, ClusterState clusterState) {
```
good move, folding this into the base class..

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/optimize/ShardOptimizeRequest.java:
```diff
@@ -20,41 +20,37 @@
     @Override
     public void readFrom(StreamInput in) throws IOException {
         super.readFrom(in);
-        request.readFrom(in);
```
these overrides can go completely now, no?

jasontedor replied:
Agree. Addressed.

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/recovery/RecoveryResponse.java:
```diff
@@ -72,7 +73,7 @@ public void detailed(boolean detailed) {
     }
 
-    public Map<String, List<ShardRecoveryResponse>> shardResponses() {
+    public Map<String, List<RecoveryState>> shardResponses() {
```
can we call this shardRecoveryStates?

jasontedor replied:
Agree. Addressed.

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/recovery/RecoveryResponse.java:
```diff
@@ -35,10 +36,10 @@
 
     private boolean detailed = false;
-    private Map<String, List<ShardRecoveryResponse>> shardResponses = new HashMap<>();
+    private Map<String, List<RecoveryState>> shardResponses = new HashMap<>();
```
can we call this shardRecoveryStates?

jasontedor replied:
Agree. Addressed.

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/recovery/ShardRecoveryResponse.java:
```diff
@@ -19,42 +19,46 @@
  * Information regarding the recovery state of a shard.
  */
-public class ShardRecoveryResponse extends BroadcastShardResponse implements ToXContent {
+public class ShardRecoveryResponse extends BaseNodeBroadcastResponse implements ToXContent {
```
can we call this RecoveryNodeResponse? (it's no longer about a single shard)

jasontedor replied:
Agree. Addressed.

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/recovery/ShardRecoveryResponse.java:
```diff
@@ -63,26 +67,37 @@ public void recoveryState(RecoveryState recoveryState) {
     @Override
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
-        recoveryState.toXContent(builder, params);
+        builder.startArray("recovery_states");
```
doesn't this add another field name that wasn't there before? Is this method used? do we need to implement toXContent? I'm looking a the implementation of RecoveryResponse

jasontedor replied:
We do not need to implement `ToXContent` here. Removed.

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/recovery/TransportRecoveryAction.java:
```diff
@@ -151,13 +125,18 @@ protected ClusterBlockException checkRequestBlock(ClusterState state, RecoveryRe
     }
 
-    static class ShardRecoveryRequest extends BroadcastShardRequest {
+    static class ShardRecoveryRequest extends BaseNodeBroadcastRequest<RecoveryRequest> {
```
I wonder if we should just in line the response as well? we don't expose it through the API anymore. Also, can we call this NodeRecoveryRequest?

jasontedor replied:
Agree on both counts. Addressed.

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/segments/TransportIndicesSegmentsAction.java:
```diff
@@ -83,55 +80,46 @@ protected ClusterBlockException checkRequestBlock(ClusterState state, IndicesSeg
+            concatenation.addAll(response.getShardSegments());
         }
-        return new IndicesSegmentResponse(shards.toArray(new ShardSegments[shards.size()]), clusterState, shardsResponses.length(), successfulShards, failedShards, shardFailures);
+        List<DefaultShardOperationFailedException> es = Lists.newArrayList();
```
Left over?

jasontedor replied:
Yes. Removed.

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/stats/ShardStats.java:
```diff
@@ -36,7 +36,7 @@
 /**
  */
-public class ShardStats extends BroadcastShardResponse implements ToXContent {
+public class ShardStats implements Streamable, ToXContent {
```
cool this is not a response any more...

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/stats/TransportIndicesStatsAction.java:
```diff
@@ -217,5 +202,53 @@ public void writeTo(StreamOutput out) throws IOException {
+            super.readFrom(in);
+            if (in.readBoolean()) {
+                int size = in.readVInt();
+                shards = Lists.newArrayListWithCapacity(size);
```
use ArrayList? one less usage for non standard code...

jasontedor replied:
Yes.

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/upgrade/get/TransportUpgradeStatusAction.java:
```diff
@@ -138,15 +128,60 @@ protected ShardUpgradeStatus shardOperation(IndexShardUpgradeStatusRequest reque
 
-    static class IndexShardUpgradeStatusRequest extends BroadcastShardRequest {
-
+    static class IndexShardUpgradeStatusRequest extends BaseNodeBroadcastRequest<UpgradeStatusRequest> {
```
calls this NodeUpgradeStatusRequest/Response?

jasontedor replied:
Agree. Addressed.

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/upgrade/get/TransportUpgradeStatusAction.java:
```diff
@@ -138,15 +128,60 @@ protected ShardUpgradeStatus shardOperation(IndexShardUpgradeStatusRequest reque
+            super.readFrom(in);
+            if (in.readBoolean()) {
+                int size = in.readVInt();
+                shards = Lists.newArrayListWithCapacity(size);
```
same comment about ArrayList.

jasontedor replied:
Yes.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/indices/BaseNodeBroadcastRequest.java:
```diff
@@ -0,0 +1,102 @@
+import java.io.IOException;
+import java.util.List;
+
+public abstract class BaseNodeBroadcastRequest<T extends IndicesLevelRequest> extends TransportRequest implements IndicesRequest {
```
can we call this `BaseBroadcastByNodeRequest`?

jasontedor replied:
Yes.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/indices/BaseNodeBroadcastRequest.java:
```diff
@@ -0,0 +1,102 @@
+        indicesLevelRequest = newRequest();
+        indicesLevelRequest.readFrom(in);
+        int size = in.readVInt();
+        shards = Lists.newArrayListWithCapacity(size);
```
same..

jasontedor replied:
Yes.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/indices/BaseNodeBroadcastResponse.java:
```diff
@@ -0,0 +1,99 @@
+import java.io.IOException;
+import java.util.List;
+
+public class BaseNodeBroadcastResponse extends TransportResponse {
```
since this is not abstract , do we want to call it base? I'm +1 to just remove Base from both response & request

jasontedor replied:
It was meant to be abstract.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/indices/IndicesLevelRequest.java:
```diff
@@ -0,0 +1,36 @@
+import org.elasticsearch.action.ActionRequest;
+import org.elasticsearch.action.support.broadcast.BroadcastRequest;
+
+public abstract class IndicesLevelRequest<T extends IndicesLevelRequest> extends BroadcastRequest<T> {
```
what does this class buy us? Can we just use BroadcastRequest?

jasontedor replied:
Removed.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/indices/IndicesLevelResponse.java:
```diff
@@ -0,0 +1,37 @@
+
+import java.util.List;
+
+public abstract class IndicesLevelResponse extends BroadcastResponse {
```
same here. Can we use BroadcastResponse?

jasontedor replied:
Removed.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/indices/TransportNodeBroadcastAction.java:
```diff
@@ -0,0 +1,375 @@
+ * @param <NodeBroadcastResponse> per-node container of shard-level responses
+ * @param <ShardOperationResult>  per-shard operation results
+ */
+public abstract class TransportNodeBroadcastAction<Request extends IndicesLevelRequest,
```
I think the name is confusing (given that we have TransportNodeActions which broadcast requests to nodes). I suggest TransportBroadcastByNodeAction, which is a better sibling imho to TransportBroadcastAction

jasontedor replied:
Agree. Addressed.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/indices/TransportNodeBroadcastAction.java:
```diff
@@ -0,0 +1,375 @@
+        this.clusterService = clusterService;
+        this.transportService = transportService;
+
+        transportNodeBroadcastAction = actionName + "[i]";
```
why `'[i]` ? can we use `[n]` like other per node actions?

jasontedor replied:
Yes.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/indices/TransportNodeBroadcastAction.java:
```diff
@@ -0,0 +1,375 @@
+    protected abstract NodeBroadcastResponse newNodeResponse(String nodeId, int totalShards, int successfulShards, List<ShardOperationResult> results, List<BroadcastShardOperationFailedException> exceptions);
+
+    /**
+     * Executes the shard-level operation. This method is called asynchronously once per shard.
```
old docs?

jasontedor replied:
Yes, good catch.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/indices/TransportNodeBroadcastAction.java:
```diff
@@ -0,0 +1,375 @@
+     * @param concreteIndices the concrete indices on which to execute the operation
+     * @return the shards on which to execute the operation
+     */
+    protected abstract GroupShardsIterator shards(ClusterState clusterState, Request request, String[] concreteIndices);
```
I think we should move away from GroupShardIterator and just use plain vanilla ShardIterator. This is just confusing and doesn't buy much.

jasontedor replied:
Agree.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/indices/TransportNodeBroadcastAction.java:
```diff
@@ -0,0 +1,375 @@
+            for (ShardIterator shardIt : shardIts) {
+                ShardRouting shard = shardIt.nextOrNull();
+                if (shard != null) {
+                    String nodeId = shard.currentNodeId();
```
when we move to ShardIterator we can have shards which are not assigned (and can't have null shards)

jasontedor replied:
ACK. Which are important so that we get the counts right.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/indices/TransportNodeBroadcastAction.java:
```diff
@@ -0,0 +1,375 @@
+            }
+        }
+
+        private void performOperation(final DiscoveryNode node, List<ShardRouting> shards, final int nodeIndex) {
```
sendOperationToNode?

jasontedor replied:
Agree. Addressed.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/indices/TransportNodeBroadcastAction.java:
```diff
@@ -0,0 +1,375 @@
+            }
+        }
+
+        protected void onOperation(DiscoveryNode node, int nodeIndex, NodeBroadcastResponse response) {
```
onNodeResponse? (operation suggests it's done here, imho, when compare to other on-prefixed namings)

jasontedor replied:
Agree. Addressed.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/indices/TransportNodeBroadcastAction.java:
```diff
@@ -0,0 +1,375 @@
+            }
+        }
+
+        protected void onFailure(DiscoveryNode node, int nodeIndex, Throwable t) {
```
onNodeFailure?

jasontedor replied:
Agree. Addressed.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/indices/TransportNodeBroadcastAction.java:
```diff
@@ -0,0 +1,375 @@
+                Response response = newResponse(request, responses, unavailableShardExceptions);
+                listener.onResponse(response);
+            } catch (Throwable t) {
+                logger.debug("failed to combine responses from nodes", t);
```
can we separate the two? the listener can also throw an exception...

jasontedor replied:
Yes. Addressed.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/indices/TransportNodeBroadcastAction.java:
```diff
@@ -0,0 +1,375 @@
+        public void messageReceived(final NodeBroadcastRequest request, TransportChannel channel) throws Exception {
+            List<ShardRouting> shards = request.getShards();
+            final int totalShards = shards.size();
+            logger.trace("executing operation [{}] on [{}] shards on node [{}]", actionName, totalShards, request.getNodeId());
```
can we log the actionName first? also, the node can go away, as the node name is part of the log prefix aleady. `[{}] executing operation on [{}] shards`

jasontedor replied:
Agree. Addressed.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/indices/TransportNodeBroadcastAction.java:
```diff
@@ -0,0 +1,375 @@
+            List<ShardRouting> shards = request.getShards();
+            final int totalShards = shards.size();
+            logger.trace("executing operation [{}] on [{}] shards on node [{}]", actionName, totalShards, request.getNodeId());
+            final Object[] shardResults = new Object[totalShards];
```
it seems this can be ShardOperationResult[] no?

jasontedor replied:
Shard-level exceptions are stored in this array as well.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/indices/TransportNodeBroadcastAction.java:
```diff
@@ -0,0 +1,375 @@
+
+        private void onShardOperation(final NodeBroadcastRequest request, final Object[] shardResults, final int shardIndex, final ShardRouting shardRouting) {
+            try {
+                logger.trace("executing operation [{}] for shard [{}] on node [{}]", actionName, shardRouting.shortSummary(), request.getNodeId());
```
same comment about log formatting (goes for the rest too..)

jasontedor replied:
Agree. All are addressed.

bleskes commented:
@jasontedor I did a review and I like it. Left a bunch of comments. It would also be great if we can manage to have unit testing for the new transport action like we have for TransportReplicationAction.

jasontedor commented:
@bleskes Thanks for the very helpful and thorough review. I've addressed all of your review comments. Of note, I added unit tests for `TransportBroadcastByNodeAction`. In the course of writing those tests, I [altered `TransportBroadcastByNodeAction`](https://github.com/jasontedor/elasticsearch/blob/9b9d859198f4cc43615f43036b210453ca0b2e43/core/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeAction.java#L104-L110) to count all shards on a failing node as contributing to failures that accumulate in the final response. Good?

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/cache/clear/ClearIndicesCacheRequest.java:
```diff
@@ -28,7 +27,7 @@
  *
  */
-public class ClearIndicesCacheRequest extends BroadcastRequest<ClearIndicesCacheRequest> {
+public class ClearIndicesCacheRequest extends org.elasticsearch.action.support.broadcast.BroadcastRequest<ClearIndicesCacheRequest> {
```
import glitch?

jasontedor replied:
Yes, from the IDE. Fixed, thanks.

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/cache/clear/ClearIndicesCacheResponse.java:
```diff
@@ -38,7 +38,7 @@
     }
 
-    ClearIndicesCacheResponse(int totalShards, int successfulShards, int failedShards, List<ShardOperationFailedException> shardFailures) {
+    ClearIndicesCacheResponse(int totalShards, int successfulShards, int failedShards, List<DefaultShardOperationFailedException> shardFailures) {
```
why this change?

jasontedor replied:
I'm not sure, but I've reverted it.

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/cache/clear/ShardClearIndicesCacheRequest.java:
```diff
@@ -41,8 +42,8 @@
-        super(shardId, request);
+    ShardClearIndicesCacheRequest(ClearIndicesCacheRequest request, List<ShardRouting> shards, String nodeId) {
+        super(nodeId, request, shards);
         queryCache = request.queryCache();
```
I know it's not related and was that way before, but I think these can be gotten from request? do we need to  have these fields stored and streamed separately?

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/optimize/OptimizeRequest.java:
```diff
@@ -36,7 +35,7 @@
  * @see OptimizeResponse
  */
-public class OptimizeRequest extends BroadcastRequest<OptimizeRequest> {
+public class OptimizeRequest extends org.elasticsearch.action.support.broadcast.BroadcastRequest<OptimizeRequest> {
```
Same import issue?

jasontedor replied:
Yes.

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/optimize/OptimizeResponse.java:
```diff
@@ -38,7 +38,7 @@
     }
 
-    OptimizeResponse(int totalShards, int successfulShards, int failedShards, List<ShardOperationFailedException> shardFailures) {
+    OptimizeResponse(int totalShards, int successfulShards, int failedShards, List<DefaultShardOperationFailedException> shardFailures) {
```
Same question... why is this needed?

jasontedor replied:
Also reverted.

jasontedor commented:
@bleskes I've pushed commits in response to your latest round of comments. Thanks again for reviewing.

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/segments/TransportIndicesSegmentsAction.java:
```diff
@@ -83,68 +80,92 @@ protected ClusterBlockException checkRequestBlock(ClusterState state, IndicesSeg
-                successfulShards++;
-            }
+    protected IndicesSegmentResponse newResponse(IndicesSegmentsRequest request, int totalShards, int successfulShards, int failedShards, List<IndexShardSegmentResponse> indexShardSegmentResponses, List<ShardOperationFailedException> shardFailures) {
+        List<ShardSegments> concatenation = Lists.newArrayList();
```
Simple ArrayList?

bleskes commented on core/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeAction.java:
```diff
@@ -0,0 +1,393 @@
+                    unavailableShardExceptions.add(
+                            new NoShardAvailableActionException(
+                                    shard.shardId(),
+                                    "[" + shard.shardId().getIndex() + "][" + shard.shardId().getId() + "] no shards available while executing " + actionName
```
nit: ShardId.toString does this :)

jasontedor replied:
Good to know. :)

bleskes commented on core/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeAction.java:
```diff
@@ -0,0 +1,393 @@
+                logger.debug("failed to execute on node [{}]: [{}]", nodeId, t);
+            }
+
+            responses.set(nodeIndex, new FailedNodeException(nodeId, "Failed node [" + nodeId + "]", t));
```
can we use compareAndSet here , make sure it's null and only then increment the response counter? This way we're protected from double handling of failure / responses (needed because of the throwable handling in L308)

jasontedor replied:
Yes, for safety, but we should do the same in `onNodeResponse` too.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeAction.java:
```diff
@@ -0,0 +1,393 @@
+
+        public void start() {
+            if (nodeIds.size() == 0) {
+                try {
```
just call onCompletion?

jasontedor replied:
Yes.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeAction.java:
```diff
@@ -0,0 +1,393 @@
+            List<ShardRouting> shards = request.getShards();
+            final int totalShards = shards.size();
+            logger.trace("[{}] executing operation on [{}] shards", actionName, totalShards);
+            final Object[] shardResults = new Object[totalShards];
```
nit: call this shardResultOrErrors? , no big deal though

jasontedor replied:
Sure. I like that better.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/indices/BaseBroadcastByNodeRequest.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    public IndicesOptions indicesOptions() {
+        return originalIndices.indicesOptions();
```
I think we can proxy this and indices() to the indicesLevelRequest ?

jasontedor replied:
Yes. Addressed.

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java:
```diff
@@ -222,6 +223,59 @@ public GroupShardsIterator allAssignedShardsGrouped(String[] indices, boolean in
+            }
+            for (IndexShardRoutingTable indexShardRoutingTable : indexRoutingTable) {
+                for (ShardRouting shardRouting : indexShardRoutingTable) {
+                    if (predicate.apply(shardRouting) || includeEmpty) {
```
I like the predicate based implementation , but what's the point of the includeEmpty variable, it just gives you all shards? I might be missing something but can't we just remove it and have an allShards(indices, includeReplicatingTarget) + a match all predicate?

jasontedor replied:
Done this way, the predicate method for a plain `ShardsIterator` isn't needed at this time. If you think I should add it anyway, I can but for now I've left it out. Otherwise, I've addressed this comment.

bleskes commented on core/src/test/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeActionTests.java:
```diff
@@ -0,0 +1,437 @@
+                .addGlobalBlock(new ClusterBlock(1, "", false, true, RestStatus.SERVICE_UNAVAILABLE, ClusterBlockLevel.ALL));
+        clusterService.setState(ClusterState.builder(clusterService.state()).blocks(block));
+        try {
+            action.new AsyncAction(request, listener).start();
```
nit: start not needed?

jasontedor replied:
Removed.

bleskes replied:
I guess something went wrong.. no biggy.

bleskes commented on core/src/test/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeActionTests.java:
```diff
@@ -0,0 +1,437 @@
+        }
+    }
+
+    public void testResultAggregation() throws ExecutionException, InterruptedException {
```
nice

jasontedor commented:
@bleskes I think that this is ready for another round of review.

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java:
```diff
@@ -222,6 +224,59 @@ public GroupShardsIterator allAssignedShardsGrouped(String[] indices, boolean in
+        }
+    };
+
+    public ShardsIterator allAssignedShards(String[] indices, boolean includeEmpty) {
```
can we add java docs to all of these?

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/recovery/TransportRecoveryAction.java:
```diff
@@ -61,84 +58,55 @@ public TransportRecoveryAction(Settings settings, ThreadPool threadPool, Cluster
-                    shardFailures = new ArrayList<>();
+    @Override
+    protected RecoveryResponse newResponse(RecoveryRequest request, int totalShards, int successfulShards, int failedShards, List<RecoveryState> responses, List<ShardOperationFailedException> shardFailures) {
+        Map<String, List<RecoveryState>> shardResponses = Maps.newHashMap();
```
this will annoy the forbidden API after rebase + squash. Heads up

jasontedor replied:
Are you referring to `Maps`? That hasn't been forbidden yet (but it will be soon).

bleskes replied:
kk. was referring to both the maps and the lists later on…

> On 28 Aug 2015, at 20:40, Jason Tedor <notifications@github.com> wrote:
> 
> In core/src/main/java/org/elasticsearch/action/admin/indices/recovery/TransportRecoveryAction.java:
> 
> >  
> > -        for (int i = 0; i < shardsResponses.length(); i++) {
> > -            Object shardResponse = shardsResponses.get(i);
> > -            if (shardResponse == null) {
> > -                // simply ignore non active shards
> > -            } else if (shardResponse instanceof BroadcastShardOperationFailedException) {
> > -                failedShards++;
> > -                if (shardFailures == null) {
> > -                    shardFailures = new ArrayList<>();
> > +    @Override
> > +    protected RecoveryResponse newResponse(RecoveryRequest request, int totalShards, int successfulShards, int failedShards, List<RecoveryState> responses, List<ShardOperationFailedException> shardFailures) {
> > +        Map<String, List<RecoveryState>> shardResponses = Maps.newHashMap();
> 
> @bleskes Are you referring to Maps? That hasn't been forbidden yet (but it will be soon).
> 
> —
> Reply to this email directly or view it on GitHub.
>

jasontedor replied:
Got it. I'll get the `Lists` during the squash/rebase dance, and the `Maps` when those are all removed from master.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeAction.java:
```diff
@@ -0,0 +1,546 @@
+
+    /**
+     * Determines the shards on which this operation will be executed on. The operation is executed once per shard
+     * iterator and unlike TransportBroadcastAction does not fall back to the next shard in the iterator on failure.
```
Now that we use a simpler shard iterator, this is no longer up to date.

jasontedor replied:
Addressed in 104b09365f0bf6839d7de533345ce453d0b173a4.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeAction.java:
```diff
@@ -0,0 +1,546 @@
+                throw requestBlockException;
+            }
+
+            logger.trace("resolving shards based on cluster state version [{}]", clusterState.version());
```
can we add the action name here?

jasontedor replied:
Yes. Addressed in e5e46589449565bbecf647f0752d45216bc18731.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeAction.java:
```diff
@@ -0,0 +1,546 @@
+        }
+
+        protected void onNodeResponse(DiscoveryNode node, int nodeIndex, NodeResponse response) {
+            logger.trace("received response from node [{}]", node.id());
```
add action name pls.

jasontedor replied:
Yes, addressed in 4b57fa400290b8fcdecdaf0b0b744fdfe72d2a94.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeAction.java:
```diff
@@ -0,0 +1,546 @@
+        protected void onNodeFailure(DiscoveryNode node, int nodeIndex, Throwable t) {
+            String nodeId = node.id();
+            if (logger.isDebugEnabled() && !(t instanceof NodeShouldNotConnectException)) {
+                logger.debug("failed to execute on node [{}]: [{}]", nodeId, t);
```
action name...

jasontedor replied:
Yes, addressed in 4b57fa400290b8fcdecdaf0b0b744fdfe72d2a94.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeAction.java:
```diff
@@ -0,0 +1,546 @@
+            out.writeString(nodeId);
+        }
+
+        public Request getIndicesLevelRequest() {
```
unused now..

jasontedor replied:
Yup. Removed in 0ea515a67e46d18ba7dbc27f7f40a1456985da17.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeAction.java:
```diff
@@ -0,0 +1,546 @@
+        }
+    }
+
+    public final static class EmptyResult implements Streamable {
```
add java docs on when to use this?

jasontedor replied:
Yes. Added in 68d0c8ecde9054e08a885a97e75b6475940179e2.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/broadcast/node/TransportBroadcastByNodeAction.java:
```diff
@@ -0,0 +1,546 @@
+        }
+    }
+
+    protected class NodeRequest extends TransportRequest implements IndicesRequest {
```
I think we don't need the indicesRequests here anymore. Can we remove it and the corresponding methods?

jasontedor replied:
Yes. Removed in 098d9cec9b733b61341e6cb58412b6dc9232944e.

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java:
```diff
@@ -222,6 +224,37 @@ public GroupShardsIterator allAssignedShardsGrouped(String[] indices, boolean in
         return new GroupShardsIterator(set);
     }
 
+    public ShardsIterator allShardsExcludingRelocationTargets(String[] indices) {
```
relocating targets are something we create on the fly - most of the time they are not needed and do not really exists in the routing table (but do in routing nodes!). I think we can just call this allShards..

jasontedor replied:
Agree. This has been renamed in d58ffadf951542af88b064b8b091c892cb1260d3.

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java:
```diff
@@ -222,6 +224,37 @@ public GroupShardsIterator allAssignedShardsGrouped(String[] indices, boolean in
+            }
+            for (IndexShardRoutingTable indexShardRoutingTable : indexRoutingTable) {
+                for (ShardRouting shardRouting : indexShardRoutingTable) {
+                    if (predicate.apply(shardRouting)) {
```
nice :)

bleskes commented on core/src/main/java/org/elasticsearch/cluster/routing/RoutingTable.java:
```diff
@@ -222,6 +224,37 @@ public GroupShardsIterator allAssignedShardsGrouped(String[] indices, boolean in
+                    if (predicate.apply(shardRouting)) {
+                        shards.add(shardRouting);
+                    }
+                    if (includeRelocationTargets && shardRouting.relocating()) {
```
Wondering if we should  only add the relocation target if the source shard matches the predicate...  WDYT?

jasontedor replied:
Yeah, good catch. That's definitely where it should be. Addressed in 125f0b738c0b44c8fec7200a5a3dd67b97eb1a1c.

bleskes commented:
LGTM. Left some minor comments - no need for another review - feel free to push once addressed. I think this has turned out pretty nice.

jasontedor commented:
Thanks @bleskes for another round of helpful reviews, especially the base class request/response simplification. I'll get this integrated into master soon.

bleskes commented:
oh - one more thing - after you squash / rebase etc. Can you make sure to add a note to the commit message and this PR that we added an option to register transport handlers and why? I case someone wonders ...

jasontedor commented:
This pull request has been merged to master (5e2efcfe091c62ef215016e2981d05339ea945a3) and integrated into the 2.0 branch (b6287bb37355a5d84639e7cb015810b6c112e088).

bleskes commented:
@jasontedor great! I just realized that TransportUpgradeAction should probably be migrated as well. Can you take care of it?

jasontedor commented:
@bleskes I've opened #13205 to address `TransportUpgradeAction`.

mikemccand commented:
In the nightly benchmarks (https://benchmarks.elastic.co/index.html ) it looks like this change caused an increase in indices stats time (see annotation B on the Stats request time graph).

I guess this is expected, since now the stats computation for the 5 shards is done in serial on the one node, instead of concurrently?

jasontedor commented:
> In the nightly benchmarks (https://benchmarks.elastic.co/index.html ) it looks like this change caused an increase in indices stats time (see annotation B on the Stats request time graph).

Thanks for bringing this to my attention @mikemccand.

> I guess this is expected, since now the stats computation for the 5 shards is done in serial on the one node, instead of concurrently?

Precisely. It looks the regression here is 2x, from ~20ms to ~40ms. If this turns out to be a major issue, we have some options to address.
